src/main/surfrdUtilsMod.F90

```diff
@@ -143,6 +145,140 @@ subroutine convert_cft_to_pft( begg, endg, cftsize, wt_cft )
+    use pftconMod, only: noveg, nc3crop, nc3irrig, pftcon
+    use array_utils, only: find_k_max_indices
+    !
+    ! !ARGUMENTS:
+    ! Use begg and endg rather than 'bounds', because bounds may not be
+    ! available yet when this is called
+    integer, intent(in) :: begg  ! Beginning grid cell index
+    integer, intent(in) :: endg  ! Ending grid cell index
+    integer, intent(in) :: natpft_size  ! how many pfts 
+    integer, intent(in) :: n_dom_pfts  ! # dominant soil pfts
+    ! These arrays modified in-place
+    ! Weights of pfts per grid cell
+    ! Dimensioned [g, natpft_lb:natpft_ub]
+    real(r8), intent(inout) :: wt_nat_patch(begg:, natpft_lb:)
+    !
+    ! !LOCAL VARIABLES:
+    integer :: g  ! gridcell index
+    integer :: m  ! pft index
+    integer :: n  ! index of the order of the dominant pfts
+    integer, allocatable :: max_indices(:)  ! array of dominant pft index values
+    real(r8) :: wt_dom_nat_sum
+
+    character(len=*), parameter :: subname = 'collapse_nat_pfts'
+
+    !-----------------------------------------------------------------------
+
+    SHR_ASSERT_ALL((ubound(wt_nat_patch) == (/endg, natpft_lb+natpft_size-1/)), errMsg(sourcefile, __LINE__))
+
+    ! Find the top N dominant pfts to collapse the data to
+    ! n_dom_pfts < 0 is not allowed (error check in controlMod.F90)
+    ! Default value n_dom_pfts = 0 or a user-selected n_dom_pfts = natpft_size
+    ! means "do not collapse pfts" and skip over this subroutine's work
+    if (n_dom_pfts > 0 .and. n_dom_pfts < natpft_size) then
+       allocate(max_indices(n_dom_pfts))
+       do g = begg, endg
+          max_indices = 0._r8  ! initialize
+          call find_k_max_indices(wt_nat_patch(g,:), natpft_lb, n_dom_pfts, &
+                                  max_indices)
+
+          ! Adjust wt_nat_patch by normalizing the dominant weights to 1
+          ! (currently they sum to <= 1) and setting the remaining weights to 0.
+          ! TODO Possible to use existing function renormalize?
+          wt_dom_nat_sum = 0._r8  ! initialize the dominant pft sum
+          do n = 1, n_dom_pfts
+             m = max_indices(n)
+             wt_dom_nat_sum = wt_nat_patch(g,m) + wt_dom_nat_sum
+          end do
+          ! Normalize dominant pft weights to 1; if non-existent,
+          ! set the pft weights and pft landunit weight to 0.
+          if (wt_dom_nat_sum <= 0._r8) then
+             wt_nat_patch(g,:) = 0._r8
+             wt_nat_patch(g,noveg) = 1._r8  ! to pass error check below
+          else
+             do n = 1, n_dom_pfts
+                m = max_indices(n)
+                wt_nat_patch(g,m) = wt_nat_patch(g,m) / wt_dom_nat_sum
+             end do
+          end if
+          ! Set non-dominant pft weights to 0
+          do m = 0, natpft_ub
+             if (.not. any(max_indices == m)) then
+                wt_nat_patch(g,m) = 0._r8
+                pftcon%is_pft_known_to_model(m) = .false.
```

Comment:
This setting of `is_pft_known_to_model` looks wrong to me - particularly since this `is_pft_known_to_model` is not gridcell-specific, but you're setting this inside a gridcell-specific loop.

Reply:
I see what you mean... I will remove that line. I have a feeling I don't need it.

Reply:
Let me reword... I will remove it because it's wrong, and I don't think that I must replace it with anything because it also serves no purpose than of being a bug :-)